src/Control/Exception/Safe.hs

```diff
@@ -344,28 +387,28 @@ tryAnyDeep = tryDeep
 -- <https://github.com/fpco/safe-exceptions#quickstart>
 --
 -- @since 0.1.0.0
-tryAsync :: (C.MonadCatch m, E.Exception e) => m a -> m (Either e a)
+tryAsync :: HAS_CALL_STACK => (C.MonadCatch m, E.Exception e) => m a -> m (Either e a)
 tryAsync = C.try
 
 -- | A variant of 'try' that takes an exception predicate to select
 -- which exceptions are caught.
 --
 -- @since 0.1.4.0
-tryJust :: (C.MonadCatch m, Exception e) => (e -> Maybe b) -> m a -> m (Either b a)
-tryJust f a = catch (Right `liftM` a) (\e -> maybe (throwM e) (return . Left) (f e))
+tryJust :: HAS_CALL_STACK => (C.MonadCatch m, Exception e) => (e -> Maybe b) -> m a -> m (Either b a)
+tryJust f a = withFrozenCallStack catch (Right `liftM` a) (\e -> maybe (throwM e) (return . Left) (f e))
 
 -- | Async safe version of 'E.onException'
 --
 -- @since 0.1.0.0
-onException :: C.MonadMask m => m a -> m b -> m a
-onException thing after = withException thing (\(_ :: SomeException) -> after)
+onException :: HAS_CALL_STACK => C.MonadMask m => m a -> m b -> m a
+onException thing after = withFrozenCallStack withException thing (\(_ :: SomeException) -> after)
 
 -- | Like 'onException', but provides the handler the thrown
 -- exception.
 --
 -- @since 0.1.0.0
-withException :: (C.MonadMask m, E.Exception e) => m a -> (e -> m b) -> m a
-withException thing after = C.uninterruptibleMask $ \restore -> do
+withException :: HAS_CALL_STACK => (C.MonadMask m, E.Exception e) => m a -> (e -> m b) -> m a
+withException thing after = withFrozenCallStack $ C.uninterruptibleMask $ \restore -> do
```

Comment:
There's a subtle thing here - `withFrozenCallStack $` is going to freeze the `CallStack` for `thing` and `after`, while `withFrozenCallStack C.uninterruptibleMask` is going to only freeze the `CallStack` for that function. I suspect we want the former, though I *also* suspect that it's going to play weirdly with the types. May be worth not freezing here.

Reply:
>  I suspect we want the former

Did you mean to say `latter` here, by any chance? If so I agree with your comment: `withFrozenCallStack C.uninterruptibleMask` seems closer to what we want (freeze the handler only), and it is consistent with other functions. And you are right about it misbehaving. Naively changing it to

```haskell
withFrozenCallStack C.uninterruptibleMask $ \restore
```
gives

```
• Couldn't match type: m a -> m a
                 with: forall a1. m a1 -> m a1
  Expected: (forall a1. m a1 -> m a1) -> m a
    Actual: (m a -> m a) -> m a
```

Reply:
Fun fact, ghc can be placated with `-XImpredicativeTypes`

```haskell
-- requires -XImpredicativeTypes
uninterruptibleMaskFrozen :: forall m b. C.MonadMask m => ((forall a. m a -> m a) -> m b) -> m b
uninterruptibleMaskFrozen = withFrozenCallStack C.uninterruptibleMask

withException thing after = uninterruptibleMaskFrozen $ \restore -> do ...
```

I'm not sure how to do this without it, however (`-XRankNTypes` is not enough).